core/src/agent/state.rs

```diff
@@ -40,10 +36,9 @@ impl AgentState {
         self.keys.clone()
     }
 
-    /// getter for a copy of self.top_pair
-    /// should be used with a source chain for validation/safety
-    pub fn top_pair(&self) -> Option<Pair> {
-        self.top_pair.clone()
+    /// getter for the chain
+    pub fn chain(&self) -> Chain {
+        self.chain.clone()
```

Comment:
Maybe not do this. This can get out of hand when used overly. better return an immutable reference to `self.chain`

Reply:
@sphinxc0re ok i'll change it, i have some questions about this and mutability but they can wait until later